jaxrs-api/src/main/java/javax/ws/rs/JAXRS.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2018 Markus KARG. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package javax.ws.rs;
+
+import java.net.URI;
+import java.util.ServiceLoader;
+
+import javax.ws.rs.core.Application;
+
+/**
+ * Bootstrap class that is used to startup a JAX-RS application in Java SE
+ * environments.
+ *
+ * The {@code JAXRS} class is available in a Jakarta EE container environment as
+ * well; however, support for the Java SE bootstrapping APIs is <em>not
+ * required</em> in container environments.
+ *
+ * @author Markus KARG (markus@headcrashing.eu)
+ *
+ * @since 2.2
+ */
+interface JAXRS {
+
+    /**
+     * Invoked in Java SE environments to start the provided application at the
+     * specified root URL.
+     *
+     * This method will not return until the JAX-RS application is terminated.
+     *
+     * @param rootURI
```

Comment:
Is it actually common to provide an `URI` to configure the endpoint? I'm not against it, I just didn't see this before.

Reply:
No, I would definitely prefer `URI` over `URL`. It's just that most API's I know of don't use `URL`/`URI` but some other way to specify all the required parameters for bootstrapping.

The most interesting question is: Which information does an implementation need for bootstrapping?
* A TCP port number
* Which interface to bind to
* Which protocol to use (HTTP/HTTPS/AJP)
* Socket Timeouts?
* TLS certificate to use for HTTPS? (just an example. Most likely not relevant for most apps)

Most if this can be represented with an `URI`. But maybe implementations would like to support advanced options. In this case the API should provide a way to pass these options.

Reply:
Well, the minimum is the root URI and the Application, obvisouly. Certainly we could add more parameters if needed. Which one are essential and supported by all vendors?

BTW, Jersey is using an URI:
```
    public static HttpServer createHttpServer(final URI uri, final ResourceConfig configuration) {
```
(from: https://github.com/jersey/jersey/blob/43c81e3f307e1d364094d0cba57f7618908e8145/containers/jdk-http/src/main/java/org/glassfish/jersey/jdkhttp/JdkHttpServerFactory.java#L89)

Reply:
So the root URI would provide us with the protocol, port and host address. And I agree that this is the minimum. Perhaps we just need some kind of map or properties parameter so that implementations can support advanced options?

Reply:
While `Properties` sounds like the vehicle of choice to my ears, it actually was invented for _persistent_ properties and has a focus on loading and storing them. `Map<Object,Object>` would be more lightweight.

Reply:
Sure, I'm also -1 for `Properties`, especially because they just support strings. A map would be much better. I just wanted to list all the options. 

Why do you prefer `Map<Object,Object>` over `Map<String,Object>`? I cannot think of a use case  for having object keys. JPA for example uses the latter and it works quite well.

Reply:
Using a URI/URL is very strange. Both are more general than necessary here are not the right abstraction. A host/port combination is standard, using 0.0.0.0 to bind all interfaces (again, something that would look very odd in a URL). We need a set of parameters to configure the service. Portability, which is the goal of this, would require defining standard parameters that every implementation must support. As I said, there's more work than meets the eye here to achieve portability and make this API useful.

Reply:
@chkal I simply do not see any need to synthetically restrict the key to `String` as it will work pretty well with any kind of Object. Unless the key must be human-readable (i. e. it solely serves the purpose of technically identifying the value in the map), `Object` is pretty sufficient. So every vendor (and the API itself) can define `static Object`s for the keys. This is, BTW, the best performance you can gain with comparisons, as it will never use character comparions.

Reply:
@spericas I agree that binding to 0.0.0.0 would be weird when using an `URI`. However, apart from that it feels elegant to use an `URI`, because it naturally includes the protocol, the port and the base path. But as mentioned above I'm not 100% sold on this idea. I never saw an URI being used in this context, so I'm still in the process to make up my mind if I like it or not. ;-)

Reply:
@mkarg Understood. I see advantages of using `Object` as you can for example use enumerations as keys.  But one disadvantage of it is that you are **forced** to use some kind of `static Object` as the key if you configure the runtime. So you cannot read values from config files, environment variables, etc anymore.

Reply:
@spericas I had chosen `URI` because Jersey uses that, see https://jersey.github.io/apidocs/latest/jersey/index.html. But I am also happy with `Map`.

@chkal The question is whether persisting config is _our_ problem? If someone wants that, he can simply write some `load` method which replaces _any_ text by predefined singletons. This is very beneficial as it would throw an "unknown configuration option" right at the time of parsing the text file, way before the app even tries to bootstrap the runtime.

Reply:
@quipsy-karg Sure, persisting isn't our problem. But we we should keep the API simple for real world use cases. And loading config property from external sources is a real world use case.